core/src/main/scala/kafka/controller/KafkaController.scala

```diff
@@ -1327,10 +1334,23 @@ class IsrChangeNotificationListener(controller: KafkaController) extends IZkChil
     }
   }
 
-  private def processUpdateNotifications(topicAndPartitions: immutable.Set[TopicAndPartition]) {
+  def propagateIsrChanges(): Unit = {
+    var isrChanges: Set[TopicAndPartition] = null
+    isrChangeTopicAndPartitionSet synchronized {
+      if (isrChangeTopicAndPartitionSet.isEmpty)
+        return
+      isrChanges = immutable.Set(isrChangeTopicAndPartitionSet.toArray:_*)
+      isrChangeTopicAndPartitionSet.clear()
+    }
     val liveBrokers: Seq[Int] = controller.controllerContext.liveOrShuttingDownBrokerIds.toSeq
-    controller.sendUpdateMetadataRequest(liveBrokers, topicAndPartitions)
-    debug("Sending MetadataRequest to Brokers:" + liveBrokers + " for TopicAndPartitions:" + topicAndPartitions)
+    debug("Sending MetadataRequest to Brokers:" + liveBrokers + " for TopicAndPartitions:" + isrChanges)
+    controller.sendUpdateMetadataRequest(liveBrokers, isrChanges)
```

Comment:
To do this, we need to hold a controller lock since sendUpdateMetadataRequest() is not thread safe.